update-format-crau/src/bin/test.rs

```diff
@@ -68,23 +78,104 @@ fn main() -> Result<(), Box<dyn Error>> {
     // this may also be the case for the InstallOperations
     // use header.translate_offset()
 
+    let mut siglen: u64 = 0;
+
     let signatures_bytes = match (manifest.signatures_offset, manifest.signatures_size) {
         (Some(sig_offset), Some(sig_size)) => {
             f.seek(SeekFrom::Start(header.translate_offset(sig_offset)))?;
 
             let mut buf = vec![0u8; sig_size as usize];
+            siglen = sig_size;
             f.read_exact(&mut buf)?;
             Some(buf.into_boxed_slice())
         }
         _ => None,
     };
 
-    let signatures = match signatures_bytes {
-        Some(ref bytes) => Some(proto::Signatures::parse_from_bytes(bytes)?),
-        None => None,
+    Ok(SigBuffer {
+        data: Box::leak(signatures_bytes.unwrap()),
+        len: siglen,
+    })
+}
+
+// parse_signature_data takes SigBuffer and public key file path.
+// Return only actual data, without version and special fields.
+fn parse_signature_data(sigbytes: &SigBuffer, pubkeyfile: &str) -> Result<&'static [u8], Box<dyn Error>> {
+    // Signatures has a container of the fields, i.e. version, data, and
+    // special fields.
+    let sigmessage = proto::Signatures::parse_from_bytes(sigbytes.data)?;
+    let siglen = sigbytes.len;
+
+    let mut sigbs: Box<[u8]> = vec![0; siglen as usize].into_boxed_slice();
+    let mut sigfound: bool = false;
+
+    // sigmessages.signatures[] has a single element in case of dev update payloads,
+    // while it could have multiple elements in case of production update payloads.
+    // For now we assume only dev update payloads are supported.
+    // Return the first valid signature, iterate into the next slot if error.
+    for sigx in &sigmessage.signatures {
+        match verify_sig_pubkey(sigx, pubkeyfile) {
+            Ok(sigxvec) => {
+                sigbs = sigxvec.into_boxed_slice();
+                sigfound = true;
+                break;
+            }
+            Err(_) => {
+                continue;
+            }
+        };
+    }
+
+    if !sigfound {
+        return Err("No signature found".into());
+    }
+
+    Ok(Box::leak(sigbs))
+}
+
+// Verify signature with public key
+fn verify_sig_pubkey(sig: &Signature, pubkeyfile: &str) -> Result<Vec<u8>, Box<dyn Error>> {
+    // The signature version is 2 in case of dev payload, and 1 or 2 in case of
+    // production update payloads. However, we do not explicitly check for a
+    // signature version, as the number could differ in some cases.
```

Comment:
```suggestion
    // The signature version is actually a numeration of the present signatures,
    // with the index starting at 2 if only one signature is present.
    // The Flatcar dev payload has only one signature but
    // the production payload has two from which only one is valid.
    // So, we see only "version 2" for dev payloads , and "version 1" and "version 2"
    //  in case of production update payloads. However, we do not explicitly check for a
    // signature version, as the number could differ in some cases.
```